### ISSUE TITLE
Make `initialize` optional?

### ISSUE THREAD

**Issue by kylefox** [github/actionview-component#195]
I'm wondering if it would be possible to remove the requirement for components to implement `initialize`. (I'd also be curious to know why this requirement was added in the first place).

Some of our components do not require parameters from the view: some components are entirely static, and others are able to encapsulate logic (accessing helpers, etc) from within the component's Ruby class.

But the most common use case for components that require no parameters is reusability. We have some components that are (conceptually) subclasses of another component. In #39 composition is suggested as an alternative to true inheritance — and I actually agree that composition is a better pattern.

However, the requirement to implement `initialize` makes composition using components kind of annoying because all of our "subclasses" are forced to implement an empty `initialize` method.

Here's an example: we have a high-level `AlertComponent` that accepts a handful of arguments. We have specific instances of alerts that we reuse, like a "Confirm your email address" alert and "Upgrade your account" alert. Using composition, the code looks like this:

### Alert Component

```ruby
class AlertComponent < ApplicationComponent
  attr_reader :title

  def initialize(title:)
    @title = title
  end
end
```

```erb
<div class="alert">
  <h2 class="alert-title"><% title %></h2>
  <div class="alert-content">
    <% content %>
  </div>
</div>
```

### Confirm Email Alert Component

```ruby
class ConfirmEmailAlertComponent < ApplicationComponent
  def initialize(*); end

  def render?
    current_user.requires_confirmation?
  end
end
```

```erb
<%= render(AlertComponent, title: 'Confirm your email') do %>
  <p>Please confirm your email address.</p>
<% end %>
```

### Upgrade Account Alert Component

```ruby
class UpgradeAccountAlertComponent < ApplicationComponent
  def initialize(*); end

  def render?
    current_user.subscription.expired?
  end
end
```

```erb
<%= render(AlertComponent, title: 'Upgrade your account') do %>
  <p>Please upgrade your account.</p>
<% end %>
```

---

You can see the last two components implement `def initialize(*); end`.

Of course, one option would be to _not_ use components for the last two examples, and instead use partials:

```erb
<!-- app/views/shared/alerts/_confirm.html.erb -->
<% if current_user.requires_confirmation? %>
  <%= render(AlertComponent, title: 'Confirm your email') do %>
    <p>Please confirm your email address.</p>
  <% end %>
<% end %>

<!-- use in some other view -->
<%= render 'shared/alerts/confirm %>
```

But this approach negates some of the benefits of using components in the first place. For example, display logic is leaking into the views (`<% if current_user.requires_confirmation? %>`). Plus, testing components is _much_ nicer than testing views 🙂And in general I think

```erb
<%= render ConfirmEmailAlertComponent %>
```

is more clear than

```erb
<%= render 'shared/alerts/confirm %>
```

**TL;DR —** would it be possible to remove the requirement for components to implement `initialize`? Or alternatively, is there a better approach to reusability across components?

**Comment by joelhawksley** [comment]
@kylefox thanks for the thorough write-up! ❤️ You're right to raise this question. 

The short answer is that @dylnclrk is working to remove this requirement, at least in Ruby 2.7+: https://github.com/github/actionview-component/pull/174. The related issue is https://github.com/github/actionview-component/issues/93.

The longer explanation is that we use the `initialize` method to derive the source file location for the component Ruby file, which is then used to look up the template(s) for the component. You can see this in action here: https://github.com/github/actionview-component/blob/master/lib/action_view/component/base.rb#L155

@tenderlove and I spent a good amount of time trying to find an alternative technique for looking up the component's template(s). I'd definitely be interested in an implementation that removes this requirement for all versions of Ruby, not just 2.7+.

**Comment by kylefox** [comment]
Thanks @joelhawksley, that makes sense. Feel free to close this issue if it's already something on the radar. (#93 didn't appear when I searched issues for "initialize" before submitting this one 🙃).

Forgive my ignorance, but is there a reason `__FILE__` can't be used to infer template locations?

**Comment by joelhawksley** [comment]
@kylefox if I remember correctly, `__FILE__` returns the current file being executed, which in this case would be the file containing `Base`, not the component that is subclassing `Base`. 

I'll close this for now, as I _believe_ the Ruby 2.7+ solution is the only viable route at this point.
